runtime/lua/vim/lsp/util.lua

```diff
@@ -1304,27 +1324,64 @@ function M.open_floating_preview(contents, syntax, opts)
     opts = { opts, 't', true };
   }
   opts = opts or {}
+  opts.fancy_markdown = opts.fancy_markdown == nil and true or opts.fancy_markdown
+  opts.close_events = opts.close_events or {"CursorMoved", "CursorMovedI", "BufHidden", "InsertCharPre"}
+
+  local bufnr = api.nvim_get_current_buf()
+
+  -- check if this popup is focusable and we need to focus
+  if opts.focus_id then
+    -- Go back to previous window if we are in a focusable one
+    local current_winnr = api.nvim_get_current_win()
+    if npcall(api.nvim_win_get_var, current_winnr, opts.focus_id) then
+      api.nvim_command("wincmd p")
+      return bufnr, current_winnr
+    end
+    do
+      local win = find_window_by_var(opts.focus_id, bufnr)
+      if win and api.nvim_win_is_valid(win) and vim.fn.pumvisible() == 0 then
+        -- focus and return the existing buf, win
+        api.nvim_set_current_win(win)
+        api.nvim_command("stopinsert")
+        return api.nvim_win_get_buf(win), win
+      end
+    end
+  end
+
+  local floating_bufnr = api.nvim_create_buf(false, true)
+  local is_fancy_markdown = syntax == "markdown" and opts.fancy_markdown
```

Comment:
I'm still not a fan of matching based on the syntax rule set (see earlier comment abou fancy_markdown)

Reply:
What do you mean exactly?

Previously, the fancy formatting was activated when the syntax was `lsp_markdown`.
I changed that to an option instead. The option only works for markdown of course.

Or would you keep a separate method for this instead of making it part of `open_floating_preview`?

What would you propose?

Reply:
I would just make it `opts.format_markdown` if true, use format_markdown, else, just display it as syntax markdown and let the gui handle it

Reply:
@mjlbach that's how it works right now no? :)

(apart from the name change to `stylize_markdown`)

Reply:
`local is_fancy_markdown = syntax == "markdown" and opts.fancy_markdown`

I'm saying don't match on syntax, match on opts.stylize markdown which defaults to true for the function

Reply:
The same function is also used for non-markdown stuff, so if we don't extra check on `syntax` we would apply the special markdown formatting for other syntaxes?

I really think we need to only call stylize markdown if what we're putting in the float is effectively markdown.

Reply:
This is the function def:

![image](https://user-images.githubusercontent.com/292349/120024518-2fa7cd00-bfa4-11eb-8cc3-30cfdabf6f77.png)

Reply:
Yeah, I guess I'm saying the function argument "syntax" seems a bit overloaded to me, as syntax "markdown"/"lsp_markdown" conditionally change a lot of paths in the code in the ways that aren't expected. I would say opts.sytlize_markdown = true + a check that markdown/lsp_markdown are the set syntax. does that make sense?

Reply:
> That is effectively what we're doing right now.

It is, but the interface is different. you should be able to pass syntax="markdown" and opts.stylize_markdown = false as a gui

Reply:
I don't think it's weird, we only call it 2-3 places so we can just pass in opts.stylize_markdown = true for hover, etc.

Reply:
> It is, but the interface is different. you should be able to pass syntax="markdown" and opts.stylize_markdown = false as a gui

And why wouldn't you be able to do that now?